modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/leases/LeaseBatchSerializer.java

```diff
@@ -0,0 +1,485 @@
+    private static final int DUMMY_LEASE_MASK = 1 << 5;
+
+    // When there are no more 8 nodes in the cluster, node name index and node index are guaranteed to fit in 7 bits we have in a varint
+    // byte, which allows us to enable 'compact mode' to save 1 byte per lease.
+
+    /** Number of bits to fit name index/node index in to enable compact mode. */
+    private static final int BIT_WIDTH_TO_FIT_IN_HALF_BYTE = 3;
+
+    /** Max size of cluster which allows compact mode. */
+    private static final int MAX_NODES_FOR_COMPACT_MODE = 1 << BIT_WIDTH_TO_FIT_IN_HALF_BYTE;
+
+    /** Mask to extract lease holder index from compact representation. */
+    private static final int COMPACT_HOLDER_INDEX_MASK = (1 << BIT_WIDTH_TO_FIT_IN_HALF_BYTE) - 1;
+
+    @Override
+    protected void writeExternalData(LeaseBatch batch, IgniteDataOutput out) throws IOException {
+        long minTimestampPhysical = minTimestampPhysicalPart(batch);
+        long commonLeasePeriod = mostFrequentLeasePeriod(batch);
+
+        out.writeVarInt(minTimestampPhysical);
+        out.writeVarInt(commonLeasePeriod);
+
+        NodesDictionary nodesDictionary = buildNodesDictionary(batch);
+        nodesDictionary.writeTo(out);
+
+        List<Lease> tableLeases = batch.leases().stream()
+                .filter(lease -> lease.replicationGroupId() instanceof TablePartitionId)
+                .collect(toList());
+        List<Lease> zoneLeases = batch.leases().stream()
+                .filter(lease -> lease.replicationGroupId() instanceof ZonePartitionId)
+                .collect(toList());
+        assert tableLeases.size() + zoneLeases.size() == batch.leases().size() : "There are " + batch.leases().size()
+                + " leases in total, "
+                + tableLeases.size() + " of them are table leases, " + zoneLeases.size() + " are zone leases, but "
+                + (batch.leases().size() - tableLeases.size() - zoneLeases.size()) + " are neither";
+
+        writePartitionedGroupLeases(tableLeases, minTimestampPhysical, commonLeasePeriod, nodesDictionary, out);
+
+        assert zoneLeases.isEmpty() : "There are zone leases which are not supported yet";
+    }
+
+    private static long minTimestampPhysicalPart(LeaseBatch batch) {
+        long min = HybridTimestamp.MAX_VALUE.getPhysical();
+
+        for (Lease lease : batch.leases()) {
+            min = Math.min(min, lease.getStartTime().getPhysical());
+            min = Math.min(min, lease.getExpirationTime().getPhysical());
+        }
+
+        return min;
+    }
+
+    private static long mostFrequentLeasePeriod(LeaseBatch batch) {
+        if (batch.leases().isEmpty()) {
+            return 0;
+        }
+
+        Long2IntMap counts = new Long2IntOpenHashMap();
+
+        for (Lease lease : batch.leases()) {
+            long period = lease.getExpirationTime().getPhysical() - lease.getStartTime().getPhysical();
```

Comment:
pls add assertion that there is no logical part

Reply:
There *can* be non-zero logical parts (they appear when prolonging). The code you pointed to is about collecting most common *absolute* parts; absolute and logical parts are encoded independently.